### PR TITLE
@intern challenges/three

### PR DESCRIPTION
**Pull request by Macbeth98**
Challenge 3: MakerDAO's Bridge Invariant

### CODE REVIEW

File: makerdao-bridge-invariant/src/L1Escrow.agent.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  public async handleTransaction(txEvent: TransactionEvent) {
+    const findings: Finding[] = [];
+
+    const logs = txEvent.filterLog(this.eventAbi);
+
+    await Promise.all(
+      logs.map(async (log) => {
+        if (log.address.toLowerCase() !== this.gatewayAddress.toLowerCase()) {
+          return;
+        }
```
Comment (Vxatz) [inline makerdao-bridge-invariant/src/L1Escrow.agent.ts]: Could have passed the `gatewayAddress` as a second argument to the `.filterLog()`

File: makerdao-bridge-invariant/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  severity: FindingSeverity;
+  type: FindingType;
+  metadata: {
+    [key: string]: string;
+  };
+};
+
+export const getNormalizedAmount = (amount: ethers.BigNumber, decimals: number) => {
+  const divValue =
+    decimals > 12 ? ethers.BigNumber.from(10 ** 12).mul(10 ** (decimals - 12)) : ethers.BigNumber.from(10 ** decimals);
```
Comment (Vxatz) [inline makerdao-bridge-invariant/src/utils.ts]: If the reason was the overflow, there's this simpler way that also works:

```suggestion
  const divValue = ethers.BigNumber.from((10 ** decimals).toString());
```

File: makerdao-bridge-invariant/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      findings = (
+        await Promise.all([
+          await arbitrumL1Escrow.handleTransaction(txEvent),
+          await optimismL1Escrow.handleTransaction(txEvent),
+        ])
+      ).flat();
+    } else {
+      const networkManager = new NetworkManager(networkData);
+
+      await networkManager.init(provider);
```
Comment (Vxatz) [inline makerdao-bridge-invariant/src/agent.ts]: This is an expensive call to be made in every transaction. You can add a `provideInitialize` block which is called only once.

File: makerdao-bridge-invariant/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+              [totalSupplyABI],
+              txEvent.blockNumber,
+              provider
+            );
+
+            const mintFinding = finding(
+              to,
+              getNormalizedAmount(value, l1DAI_Decimals).toString(),
+              totalSupply.toString()
+            );
```
Comment (Vxatz) [inline makerdao-bridge-invariant/src/agent.ts:101]: I believe this Info alert on L2s is redundant.

**Review by Vxatz — COMMENTED** [review]
Great job 👍 

I would say that you made things harder than they were supposed to be.
If you'd use `handleBlock` instead of the `handleTransaction` and query the balance on L1  once per block firing alerts when there's a change compared to the previous block (then monitoring total supply on L2 if there was an L1 alert), then the logic would be much more simplified and there would be no need of monitoring the Deposit Events.

I believe this way you should also monitor withdrawals from the L1 escrow/gateway, as these would be more likely to result in a violation of the invariant.

File: makerdao-bridge-invariant/src/utils.ts (hunk 2 above)
Comment (Macbeth98) [reply makerdao-bridge-invariant/src/utils.ts]: Oh Yes, I could have just done .toString()!!

File: makerdao-bridge-invariant/src/L1Escrow.agent.ts (hunk 1 above)
Comment (Macbeth98) [reply makerdao-bridge-invariant/src/L1Escrow.agent.ts]: Yes, I have updated my code with that

**Comment by Macbeth98** [comment]
> 



> If you'd use `handleBlock` instead of the `handleTransaction` and query the balance on L1 once per block firing alerts when there's a change compared to the previous block (then monitoring total supply on L2 if there was an L1 alert), then the logic would be much more simplified and there would be no need of monitoring the Deposit Events.

Hah yes, I realized it now after seeing your comment. I also thought of doing it per Block, but I thought I might have to give alert for each block and check balance on L2 for each block. I didn't have this idea of checking the balance with the previous block. Because of this, I thought the balance will change when there is a deposit, So I tracked the deposits. But this is a very efficient method, I didn't have the idea to check the previous block. 

Thank you for the review.  Actually, when I get time, I will create a new project with the approach you suggested and for this one I will add the monitoring of withdrawals also.

**Pull request merged**